### PR TITLE
fix: Better support for `imageSrcSet` + `imageSizes` in route `links` (#184)

### PR DESCRIPTION
**Pull request by chaance**
This PR improves support for returned links that include an `imagesrcset` attribute to preload images when matched against a responsive set of URLs. This only works when `rel === 'preload'` and `as === 'image'`.  https://html.spec.whatwg.org/commit-snapshots/cb4f5ff75de5f4cbd7013c4abad02f21c77d4d1c/#attr-link-imagesrcset

Prior to this PR, a few problems with our implementation:

1. Our types for `HtmlLinkDescriptor` make `href` required. This is true *unless* a link includes `imagesrcset` and `imagesizes`, though `imagesrcset` and `href` will also work with or without `imagesizes` according to the spec. This PR uses an intersection type to account for these cases.
2. React 17 was released prior to these attributes being accepted into the spec and implemented by browsers, so passing the props as camel-cased will throw a warning about unsupported DOM props. Our types recognized this and supported the all-lowercase version to skip the warning and forward the attribute to the node, but React 18 now recognizes the camel-cased version. This PR updates our types to use the camel-cased prop names, but it will support either at runtime and normalize the object based on the detected version of React.
3. Because we use `href` + `rel` as a key, links without an href will have duplicate keys. So `imageSrcSet` is added as a fallback to prevent that.

---

Closes: #184

- [ ] Docs
- [x] Tests

### CODE REVIEW

**Comment by ryanflorence** [comment]
I’ve found these kinds of tests to be flakey because the browser can decide not to prefetch given different memory/cpu/who-knows-what.  Maybe we can assert on the rendered link tags instead of requests?

**Comment by ryanflorence** [comment]
Also, maybe we can just use the DOM types here?

**Comment by chaance** [comment]
> Also, maybe we can just use the DOM types here?

Unfortunately the DOM types also require `href` so this wouldn't really fix things for our end users, and React's `LinkHTMLAttributes` type is different for 17 vs. 18. I think our custom types are a big DX improvement for both, personally.

File: packages/remix-react/links.ts [hunk 1]
```diff
@@ -91,16 +86,6 @@ export interface HtmlLinkDescriptor {
...
-  /**
-   * Images to use in different situations, e.g., high-resolution displays, small monitors, etc. (for rel="preload")
-   */
-  imagesrcset?: string;
-
-  /**
-   * Image sizes for different page layouts (for rel="preload")
-   */
-  imagesizes?: string;
-
```
Comment (MichaelDeBoey) [inline packages/remix-react/links.ts:103]: Removing this will technically be a breaking change I think 🤔😕

Comment (chaance) [reply packages/remix-react/links.ts:103]: Yeah, I can add these back and just mark them as `@deprecated`

Comment (MichaelDeBoey) [reply packages/remix-react/links.ts:103]: We can remove them in Remix v2 👍

Comment (chaance) [reply packages/remix-react/links.ts:103]: Appended here so they are easy to remove later https://github.com/remix-run/remix/blob/6e5a89c97a5a67bcd135cce88c9d0787ec7b3e08/packages/remix-react/links.ts#L186

File: packages/remix-server-runtime/links.ts [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
- * Represents a `<link>` element.
- *
- * WHATWG Specification: https://html.spec.whatwg.org/multipage/semantics.html#the-link-element
- */
-export interface HtmlLinkDescriptor {
+type Primitive = null | undefined | string | number | boolean | symbol | bigint;
+
+type LiteralUnion<LiteralType, BaseType extends Primitive> =
+  | LiteralType
+  | (BaseType & Record<never, never>);
```
Comment (MichaelDeBoey) [inline packages/remix-server-runtime/links.ts:5]: Is there a reason why we're not using `type-fest` here? 🤔

Comment (mcansh) [reply packages/remix-server-runtime/links.ts:5]: at the time the only type we used from type-fest was this, but i think we actually have the dep now anyways

https://github.com/remix-run/remix/pull/1123#issuecomment-1035921934

Comment (MichaelDeBoey) [reply packages/remix-server-runtime/links.ts:5]: Just looked into this & we have `type-fest` in `@remix-run/dev`, but it is in `devDependencies` though.

For `@remix-run/server-runtime` it should be in `dependencies` as we're exporting the types.

Comment (chaance) [reply packages/remix-server-runtime/links.ts:5]: May as well just inline it. No reason to worry about carrying over the dependency for this IMO.

**Review by MichaelDeBoey — APPROVED** [review]

**Pull request merged**